File: Data Structures & Algorithms/combination-target-sum-ii/submission-3.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        counter = collections.Counter(candidates)
        counter_keys = list(counter.keys())
        counter_keys.sort()
        n = len(counter_keys)
        ret = []

        def backtrack(i, curr, result):
            if curr == target:
                ret.append(result[:])
                return
            if i == n:
                return
            backtrack(i+1, curr, result)
            candidate = counter_keys[i]
            for j in range(counter[candidate]):
                curr += candidate
                if curr > target: break
                backtrack(i+1, curr, result+[candidate]*(j+1))
            
        backtrack(0, 0, [])
        return ret

File: Data Structures & Algorithms/three-integer-sum/submission-3.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ret;
        sort(nums.begin(), nums.end());
        for (int i=0; i<nums.size()-2; i++) {
            int j = i+1;
            int k = nums.size()-1;
            if (i>0 && nums[i] == nums[i-1]) continue;
            while (j<k) {
                if (j>i+1 && nums[j] == nums[j-1]) {
                    j++;
                    continue;
                }
                if (k<nums.size()-1 && nums[k] == nums[k+1]) {
                    k--;
                    continue;
                }
                int sum = nums[i] + nums[j] + nums[k];
                if (sum>0) k--;
                else if (sum<0) j++;
                else {
                    ret.push_back({nums[i], nums[j], nums[k]});
                    j++;
                    k--;
                }
            } 
        }
        return ret;
    }
};
